presto-native-execution/presto_cpp/main/operators/StorageBasedLocalShuffle.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "presto_cpp/main/operators/ShuffleInterface.h"
+#include "velox/buffer/Buffer.h"
+#include "velox/common/file/File.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/common/memory/Memory.h"
+
+namespace facebook::presto::operators {
+
+/// This class implements a persistent shuffle server that extends
+/// ShuffleInterface for read and write and also uses generalized Velox
+/// file system to maintain its state and data.
+///
+/// Except for in-progress blocks of current output vectors in the writer,
+/// each produced vector is stored as a binary file of unsafe rows. Each block
+/// filename reflects the partition and sequence number of the block (vector)
+/// for that partition. For example <ROOT_PATH>/10_12.bin is the 13th (block)
+/// vector in partition #10.
+///
+/// The class uses filesystem also to figure out the number of written shuffle
+/// files for each partition. This allows this class to be usable under
+/// different multi-threaded or multi-process scenarios as long as each producer
+/// or consumer is assigned to a distinct group of partition IDs. Each of them
+/// can create an instance of this class (pointing to the same root path) and
+/// use it to read and write shuffle data.
+class StorageBasedLocalShuffle : public ShuffleInterface {
```

Comment:
Maybe rename this to TestingPersistentShuffle. Testing clarifies that this should be used only for testing. Persistent is a shorter version of Storage-Based.